### PR TITLE
Randomized Algorithms for Optimizing Large Join Queries

### PR DESCRIPTION
**Pull request by SHSenpai**
I've edited a bit in the base class, which I want to use for all three algorithms.
Because of this, I've decided to create this is as my common branch. To make things easier for you to know what is what and keep order, I will create branches for each of the algorithms _out of this branch_. I'm also going to start working on two-phase optimization (2PO) soon, and I'll likely look back and forth a lot between all of the algorithms as I want to possibly have the Iterative and Annealing algorithms have an optimize function which is static and ideal for use in the 2PO in a way that the 2PO becomes sufficiently usable with different settings (input values to the 2PO will include things such as how many generations II will run before SA takes over, etc).

### CODE REVIEW

**Comment by SHSenpai** [comment]
> That makes it very hard for me to see what you have done here (at least if I look only in the Github Web interface, which is where I have to look to put review comments).

Gotcha. This branch contains all changes to SA, as well as the introduction of a generic function to the base class which I use in the base class.

The SA class is almost but not quite done. There's something in the paper which I simply do not understand, which I need to understand before I can finish the algorithm. The algorithm paper implies that 16*(number of joins in query) is a boolean and I don't understand what this entails.

> Simulated Annealing almost done, but I am at a loss for how to implement Equilibrium. It's supposed to be a boolean from what I can read in the paper, but the paper's example implementation defines it to be 16*(number or joins in query). (table 2.2, page 314 in the paper, page 3 in the pdf)

File: src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/randomized/SimulatedAnnealing.java
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/randomized/SimulatedAnnealing.java:47]: No, don't create such a new constructor -- see [my earlier comment](https://github.com/LiUSemWeb/HeFQUIN/pull/134#issuecomment-991719903).

However, here's what you can do: Please add the following method to `PhysicalPlanWithCostUtils`:
```java
    public static PhysicalPlanWithCost annotatePlanWithCost( final CostModel cm, final PhysicalPlan plan ) throws QueryOptimizationException {
        final Double[] costs;
        try {
            costs = CostEstimationUtils.getEstimates(cm, plan);
        } catch ( final CostEstimationException e ) {
            throw new QueryOptimizationException( "Determining the cost for the plan caused an exception.", e.getCause() );
        }

        return new PhysicalPlanWithCost( plan, costs[0] );
    }
```
Once you have that, you can change the line here as follows:
```suggestion
				final PhysicalPlanWithCost alternativePlan = PhysicalPlanWithCostUtils.annotatePlanWithCost( context.getCostModel(), temporaryPlan );
```

Comment (hartig) [reply src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/randomized/SimulatedAnnealing.java:47]: In fact, you can then also use this new helper function in line 21 above.

File: src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/randomized/SimulatedAnnealing.java
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/randomized/SimulatedAnnealing.java:49]:
```suggestion
				final double costDelta = currentPlan.getWeight() - alternativePlan.getWeight();
```

File: src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/randomized/SimulatedAnnealing.java
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/randomized/SimulatedAnnealing.java:18]:
```suggestion
	public PhysicalPlan optimize( final PhysicalPlan initialPlan ) throws QueryOptimizationException {
```

File: src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/randomized/SimulatedAnnealing.java
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/randomized/SimulatedAnnealing.java:77]:
```suggestion
	protected boolean isFrozen( final double temperature, final int unchanged ) {
```

**Comment by hartig** [comment]
> but I am at a loss for how to implement Equilibrium. It's supposed to be a boolean from what I can read in the paper, but the paper's example implementation defines it to be 16*(number or joins in query). (table 2.2, page 314 in the paper, page 3 in the pdf)

I agree that this aspect of the algorithm should have been stated more explicitly in the paper. However, my interpretation is as follows: The equilibrium is assumed to be reached after the inner loop has been run a fixed number of times, where this number is 16 * (the number of joins in the query). Hence, for queries with more joins, the inner loop runs more often.

The follow-up question then is: What do we consider as "the number of joins in the query"? The issue is that the paper considers only join plans but, in our case, there may also be other operators in the plans (i.e., not only joins). A possible solution would be to adapt the equilibrium condition to our setting as follows. Instead of fixing the number of inner-loop runs to be 16 * (the number of joins in the query), we fix it to be 16 * (the number of logical operators in the given logical plan). What do you think of this idea?

**Review by hartig — CHANGES_REQUESTED** [review]
Besides my response to your question about the equilibrium, here is my review of the current code in this PR.

File: src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/randomized/IterativeImprovementBasedQueryOptimizer.java
Comment (SHSenpai) [reply src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/randomized/IterativeImprovementBasedQueryOptimizer.java:13]: The II branch was a branch out of this branch. I wanted to make a separate branch for the IterativeImprovement changes since you wanted the PR's to be more limited in scope, but I wanted to use something from a change I had made to the base class.

Comment (hartig) [reply src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/randomized/IterativeImprovementBasedQueryOptimizer.java:13]: Okay.

**Comment by SHSenpai** [comment]
> Instead of fixing the number of inner-loop runs to be 16 * (the number of joins in the query), we fix it to be 16 * (the number of logical operators in the given logical plan). What do you think of this idea?

This seems sound, at least for now when we do not have any other idea. Testing later on will provide the data we need to evaluate the performance of all of our settings.
I'm thinking of whether I should be implementing an equilibrium condition for the SA algorithm, similar to how the II algorithm has a stopping condition, and also how I should best pass the number of operators to the optimize method, as the optimize method itself only deals with the physical plan and not the logical plan. I don't know what difference it would make, but we could also count operations in the physical plan instead of counting them in the logical plan.

Regardless, I've looked into LogicalPlanUtils, LogicalPlanVisitor, and LogicalPlanWalker to see if there's any method implemented which counts operations - as if such a method already exists I wouldn't want to reimplement the wheel, but it doesn't _seem_ like there is one.
If there isn't such a method to count operations, then should I implement it as a member for the Randomized algorithm or should I add it to the LogicalPlanUtils or another package? I'm assuming that a logical plan and its subplans is a tree without cycles, so that I can visit a subplan to see its subplans, and so on, and in that way count the number of operators (are all operators binary?).

**Comment by hartig** [comment]
> I'm thinking of whether I should be implementing an equilibrium condition for the SA algorithm, similar to how the II algorithm has a stopping condition, and also how I should best pass the number of operators to the optimize method, as the optimize method itself only deals with the physical plan and not the logical plan.

Yeah, good idea!

> I don't know what difference it would make, but we could also count operations in the physical plan instead of counting them in the logical plan.

It would not make much difference, assuming you would always be using only the initial physical plan. Other physical plans that are semantically equivalent to the initial one (i.e., "neighbors" and "neighbors' neighbors") may have different numbers of operators. Therefore, to have a constant, the initial physical plan would need to be used. However, getting this number based on the logical plan may be easier in our current implementation because of the visitor pattern that is already implemented for them, as you have found ...

> Regardless, I've looked into LogicalPlanUtils, LogicalPlanVisitor, and LogicalPlanWalker to see if there's any method implemented which counts operations - as if such a method already exists I wouldn't want to reimplement the wheel, but it doesn't seem like there is one.

Good thinking. Thanks for taking initiative!

You are right, we do not yet have such a counting functionality.

> If there isn't such a method to count operations, then should I implement it as a member for the Randomized algorithm or should I add it to the LogicalPlanUtils or another package?

LogicalPlanUtils would be the best place to put this. So, similar to the SourceAssignmentChecker that is already there, you add another class that implements LogicalPlanVisitor and that simply maintains a counter which it increases by one in all of its visit methods. Then, you add another method similar to `isSourceAssignment` and in that method you pick out the counter in the end from the visitor.

> I'm assuming that a logical plan and its subplans is a tree without cycles, so that I can visit a subplan to see its subplans, and so on, and in that way count the number of operators (are all operators binary?).

Exactly.

**Review by hartig — COMMENTED** [review]
Just one comment on the code at this stage. Everything else looks good and you can continue as discussed in [my recent comment](https://github.com/LiUSemWeb/HeFQUIN/pull/127#issuecomment-993344654).

**Comment by SHSenpai** [comment]
I've ran into a bit of a conundrum.
The equilibrium condition is based on a scaling factor, as in, the equilibrium is a generation count of x times the number of subplans in a logical plan.

The issue is where to store the amount of subplans.
If the number of subplans is to be found when the equilibrium condition is constructed and stored in the condition, then you essentially need a different condition for each plan, and therefore the same optimizer object cannot be used well for different plans. Therefore we must exclude this from consideration.
If the number of subplans is to be found when optimize is ran, we must figure this out when optimize(logicalplan) is ran and then pass it on to a new optimize(physicalplan, numberOfSubplans) method. This would preclude SimulatedAnnealing from implementing the base optimize(physicalplan) method found in RandomizedQueryOptimizerBase.
We could decide this in optimize(PhysicalPlan) by finding a way to get the number of subplans from a physical plan rather than a logical plan, but I do not know how efficient this would be.
We could remove the optimize(PhysicalPlan) method and add its functionality into optimize(LogicalPlan) since optimize(LogicalPlan) just calls optimize(PhysicalPlan) anyway, but since the optimize(PhysicalPlan) method is public I assume you want some means of working directly on a PhysicalPlan from the outside.

What do you think? @hartig

**Comment by hartig** [comment]
Yes, SimulatedAnnealing should implement optimize(logicalplan) by calling a new optimize(physicalplan, numberOfSubplans) method. Hence, it turns out that implementing optimize(logicalplan) in the abstract base class is too generic and does not work.

**Comment by SHSenpai** [comment]
There seems to be something done with the base class in main, I'll leave it to you to resolve the conflict since I'm not clued in to what you've been using the base class for elsewhere.

Since SimulatedAnnealing no longer implements optimize(physicalPlan) but instead has optimize(physicalPlan, numberOfSubplans), the base class no longer declares an abstract optimize(physicalPlan). It's commented out for now.

SA is ready now and II still succeeds with the test. I will create a new branch for 2PO as soon as you've approved the merge.

Overall it's been more work than initially expected as helper methods needed to be made and so on, but it's been really fun to learn the IDE and the language, and I'm looking forward to seeing what you have in store for me once 2PO is completed at the end of this year/start of the next. Merry Christmas, by the way!

File: src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/randomized/EquilibriumConditionByRelativeSubplanCount.java
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/randomized/EquilibriumConditionByRelativeSubplanCount.java:4]:
```suggestion
public class EquilibriumConditionByRelativeSubplanCount implements EquilibriumConditionForSimulatedAnnealing
{	
```

File: src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/randomized/EquilibriumConditionByRelativeSubplanCount.java
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/randomized/EquilibriumConditionByRelativeSubplanCount.java:9]:
```suggestion
	protected final int scalingFactor;

	public EquilibriumConditionByRelativeSubplanCount( final int scalingFactor ) {
		this.scalingFactor = scalingFactor;
	}
```

File: src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/randomized/EquilibriumConditionForSimulatedAnnealing.java
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/randomized/EquilibriumConditionForSimulatedAnnealing.java:4]:
```suggestion
public interface EquilibriumConditionForSimulatedAnnealing
{
	boolean isEquilibrium(int currentGeneration, int subplanCount);
```

File: src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/randomized/IterativeImprovementBasedQueryOptimizer.java
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/randomized/IterativeImprovementBasedQueryOptimizer.java:35]:
```suggestion
		PhysicalPlanWithCost bestPlan = PhysicalPlanWithCostUtils.annotatePlanWithCost( context.getCostModel(), initialPlan );

```

File: src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/randomized/IterativeImprovementBasedQueryOptimizer.java
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/randomized/IterativeImprovementBasedQueryOptimizer.java:42]:
```suggestion
			PhysicalPlanWithCost currentPlan = PhysicalPlanWithCostUtils.annotatePlanWithCost( context.getCostModel(), initialPlan ); // This variable will hold the plan which is currently being worked on.

```

File: src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/randomized/RandomizedQueryOptimizerBase.java
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/randomized/RandomizedQueryOptimizerBase.java:33]:
```suggestion
```

File: src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/randomized/RandomizedQueryOptimizerBase.java
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/randomized/RandomizedQueryOptimizerBase.java:35]:
```suggestion
```

File: src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/randomized/RandomizedQueryOptimizerBase.java
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/randomized/RandomizedQueryOptimizerBase.java:48]:
```suggestion
	protected <T> T getRandomElement( final List<T> L ) {
```

File: src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/randomized/SimulatedAnnealing.java
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/randomized/SimulatedAnnealing.java:27]:
```suggestion
		final PhysicalPlan initialPP = context.getLogicalToPhysicalPlanConverter().convert(initialPlan,false);
		final int numberOfSubplans = LogicalPlanUtils.countSubplans(initialPlan);
		return optimize(initialPP, numberOfSubplans);
```

File: src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/randomized/SimulatedAnnealing.java
Comment (hartig) [reply src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/randomized/SimulatedAnnealing.java:33]:
```suggestion
		PhysicalPlanWithCost bestPlan = PhysicalPlanWithCostUtils.annotatePlanWithCost( context.getCostModel(), initialPlan );
```

**Review by hartig — APPROVED** [review]

File: src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/randomized/IterativeImprovementBasedQueryOptimizer.java
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/randomized/IterativeImprovementBasedQueryOptimizer.java:28]:
```suggestion
	// @Override  TODO: fix after merging to main!!
```

File: src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/randomized/SimulatedAnnealing.java
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/randomized/SimulatedAnnealing.java:25]:
```suggestion
	// @Override TODO: fix after merging to main!!
```

**Comment by hartig** [comment]
I have merged this PR now.

Two things are remaining to complete the work on implementing the paper:
1. Change the existing code of `IterativeImprovementBasedQueryOptimizer` to use the randomized join plan optimizer.
2. Implement the third algorithm (2PO).

I suggest you do 1 next. To this end, notice that `RandomizedJoinPlanOptimizerImpl` is only for subplans that consist solely of binary joins. Therefore, you need to create a `SimpleJoinOrderingQueryOptimizer` with a `RandomizedJoinPlanOptimizerImpl` as its join plan optimizer.

**Pull request merged**
